src/executor/sort_agg.rs

```diff
@@ -0,0 +1,251 @@
+// Copyright 2022 RisingLight Project Authors. Licensed under Apache-2.0.
+#![allow(dead_code)]
+use smallvec::SmallVec;
+
+use super::*;
+use crate::array::{ArrayBuilderImpl, ArrayImpl};
+use crate::binder::BoundAggCall;
+use crate::types::{DataTypeExt, DataTypeKind};
+
+pub struct SortAggExecutor {
+    pub agg_calls: Vec<BoundAggCall>,
+    pub group_keys: Vec<BoundExpr>,
+    pub child: BoxedExecutor,
+}
+
+impl SortAggExecutor {
+    #[try_stream(boxed, ok = DataChunk, error = ExecutorError)]
+    pub async fn execute(self) {
+        let mut last_key = None::<HashKey>;
+        let mut states = create_agg_states(&self.agg_calls);
+
+        #[for_await]
+        for chunk in self.child {
+            // Eval group keys and arguments
+            let chunk = chunk?;
+            let exprs: SmallVec<[ArrayImpl; 16]> = self
+                .agg_calls
+                .iter()
+                .map(|agg| agg.args[0].eval(&chunk))
+                .try_collect()?;
+            let group_cols: SmallVec<[ArrayImpl; 16]> = self
+                .group_keys
+                .iter()
+                .map(|e| e.eval(&chunk))
+                .try_collect()?;
+
+            let num_rows = chunk.cardinality();
+            for row_idx in 0..num_rows {
+                // Create group key
+                let mut group_key = HashKey::new();
+                for col in group_cols.iter() {
+                    group_key.push(col.get(row_idx));
```

Comment:
Maybe we can optimize this in the future. We can first check group key is the same as the current group key, and then clone by `.get` if different. Otherwise we will always need to create new vector for every row of hash key.